tests/test_otel.py

```diff
@@ -0,0 +1,367 @@
+def nresponses(logger, http_ver, scheme, path, headers, n):
+    out = []
+    for _ in range(n):
+        out.append(get_response(logger, http_ver, scheme, path, headers))
+    return out
+
+
+@pytest.mark.usefixtures("trace_service", "otelcol", "nginx")
+@pytest.mark.parametrize(
+    ("http_ver", "scheme"),
+    [
+        (0, "http"),
+        (1, "http"),
+        (2, "https"),
+        (3, "https"),
+    ],
+    ids=["http 0.9", "http 1.1", "http 2.0 ssl", "http 3.0 quic"],
+)
+@pytest.mark.parametrize(
+    ("path", "headers", "text", "status"),
+    [
+        ("/", None, "TRACE-ON", 200),
+        ("/404", None, "404 Not Found", 404),
+    ],
+    ids=["ok", "error"],
+)
+def test_response_and_span_attributes(
+    http_ver, scheme, path, text, status, response, span
+):
+    assert text in (response.text if http_ver else response)
+    if http_ver:
+        assert response.status_code == status
+
+    assert span.name == ("default_location" if path == "/" else "/404")
+
+    # Default span attributes
+    assert get_attr(span, "http.method", "string_value") == "GET"
+    assert get_attr(span, "http.target", "string_value") == path
+    assert get_attr(span, "http.route", "string_value") == path
+    assert get_attr(span, "http.scheme", "string_value") == scheme
+    assert get_attr(span, "http.flavor", "string_value") == (
+        [None, "1.1", "2.0", "3.0"][http_ver]
+    )
+    assert get_attr(span, "http.user_agent", "string_value") == (
+        ([None] + [f"niquests/{niquests.__version__}"] * 3)[http_ver]
+    )
+    assert get_attr(span, "http.request_content_length", "int_value") == 0
+    assert get_attr(span, "http.response_content_length", "int_value") == (
+        len(response.text if http_ver else response)
+    )
+    assert get_attr(span, "http.status_code", "int_value") == status
+    assert get_attr(span, "net.host.name", "string_value") == "localhost"
+    assert get_attr(span, "net.host.port", "int_value") == (
+        [18080, 18080, 18443, 18443][http_ver]
+    )
+    assert get_attr(span, "net.sock.peer.addr", "string_value") == "127.0.0.1"
+    assert get_attr(span, "net.sock.peer.port", "int_value") in range(
+        1024, 65536
+    )
+
+    # Custom span attributes
+    assert get_attr(span, "http.request.completion", "string_value") == (
+        "OK" if path == "/" else None
+    )
+    assert (
+        get_attr(span, "http.response.header.content.type", "array_value")
+        .values[0]
+        .string_value
+        if path == "/"
+        else get_attr(span, "http.response.header.content.type", "array_value")
+    ) == ("text/plain" if path == "/" else None)
+    assert get_attr(span, "http.request", "string_value") == (
+        "GET /" + ["", " HTTP/1.1", " HTTP/2.0", " HTTP/3.0"][http_ver]
+        if path == "/"
+        else None
+    )
+
+
+@pytest.mark.usefixtures("trace_service", "otelcol", "nginx")
+@pytest.mark.parametrize(
+    ("http_ver", "scheme"), [(1, "http")], ids=["http 1.1"]
```

Comment:
These should be the defaults in HTTP client code, so that there is no need to create class to specify them.